src/util/expr_iterator.h

```diff
@@ -0,0 +1,247 @@
+//
+// Non-const iterators dereference to const exprt (for use with STL
+// algorithms) but have an extra .mutate() method. That method is used
+// to access non-const exprt reference but is an expensive operation
+
+/// Naive depth-first-search iterator with .mutate method for
+/// modifying exprt under the iterator.
+class depth_iteratort;
+/// Naive depth-first-search iterator.
+class const_depth_iteratort;
+/// Depth first-search iterator with duplicate avoidance (skips over
+/// duplicates, shows only first encountered child)
+/// Slower iteration than naive version. Very expensive copy.
+class const_unique_depth_iteratort;
+
+/// Depth first search iterator base - iterates over supplied expression
+/// and all its operands recursively.
+/// Base class using CRTP
+/// Do not create stand-alone instances of or upcast to this base
+/// class. It's only purpose is to provide common functions for other
+/// DFS iterators
+/// Do override .push_expr method of this class, but when doing so
+/// make this class a friend so it has acess to that overriden .push_expr
+/// method in the child class
+template<typename depth_iterator_t>
+class depth_iterator_baset:
+public std::iterator<std::forward_iterator_tag, const exprt, void>
+{
+public:
+  bool operator==(const depth_iterator_t &other) const
+  {
+    return m_stack.size()==other.m_stack.size() &&
+           std::equal(m_stack.begin(),
+                      m_stack.end(),
+                      other.m_stack.begin());
+  }
+
+  bool operator!=(const depth_iterator_t &other) const
+  {
+    return !(*this==other);
+  }
+
+  /// Preincrement operator
+  /// Do not call on the end() iterator
+  depth_iterator_t &operator++()
+  {
+    PRECONDITION(!m_stack.empty());
+    while(true)
+    {
+      if(m_stack.back().it==m_stack.back().end)
+      {
+        m_stack.pop_back();
+        if(m_stack.empty())
+        {
+          break;
+        }
+      }
+      else if(this->downcast().push_expr(*m_stack.back().it))
+      {
+        break;
+      }
+      m_stack.back().it++;
+    }
+    return this->downcast();
+  }
+
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt &operator*() const
+  {
+    PRECONDITION(!m_stack.empty());
+    return m_stack.back().expr.get();
+  }
+
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt *operator->() const
+  {
+    return &**this;
+  }
+
+protected:
+  /// Create end iterator
+  depth_iterator_baset() { }
+
+  /// Create begin iterator, starting at the supplied node
+  explicit depth_iterator_baset(const exprt &root)
+  {
+    this->downcast().push_expr(root);
+  }
+
+  /// Obtain non-const exprt reference. Breaks the tree as a side
+  /// effect. To be called only if a non-const exprt was passed into
+  /// the constructor Do not call on the end() iterator
+  exprt &mutate()
+  {
+    PRECONDITION(!m_stack.empty());
+    auto expr=std::ref(const_cast<exprt&>(m_stack.front().expr.get()));
```

Comment:
What do you mean? There's no variable or name to put & against.

The rule says: `exprt &var` instead of `exprt& var`, but `static_cast<exprt &>` seems a bit eccentric.